File: tests/fixtures/sample.pl
#!/usr/bin/perl
use strict;
use warnings;

use File::Path qw(make_path);
use Carp qw(croak);

# Maximum number of retries.
our $MAX_RETRIES = 3;
# Default port for connections.
our $DEFAULT_PORT = 8080;

# Logs a message with the given level.
sub log_message {
    my ($level, $message) = @_;
    print "[$level] $message\n";
}

package Connection;

# Creates a new Connection object.
sub new {
    my ($class, %args) = @_;
    my $self = bless {
        host      => $args{host},
        port      => $args{port} || $main::DEFAULT_PORT,
        connected => 0,
    }, $class;
    return $self;
}

# Connects to the remote host.
sub connect {
    my ($self) = @_;
    main::log_message("INFO", "Connecting to $self->{host}:$self->{port}");
    $self->{connected} = 1;
    return 1;
}

# Disconnects from the remote host.
sub disconnect {
    my ($self) = @_;
    $self->{connected} = 0;
}

# Checks if the connection is active.
sub is_connected {
    my ($self) = @_;
    return $self->{connected};
}

package Pool;

# Creates a new Pool.
sub new {
    my ($class, %args) = @_;
    my $self = bless {
        host        => $args{host},
        size        => $args{size} || 10,
        connections => [],
    }, $class;
    return $self;
}

# Acquires a connection from the pool.
sub acquire {
    my ($self) = @_;
    if (@{$self->{connections}}) {
        return pop @{$self->{connections}};
    }
    my $conn = Connection->new(host => $self->{host});
    $conn->connect();
    return $conn;
}

sub release {
    my ($self, $conn) = @_;
    push @{$self->{connections}}, $conn;
}

package main;

sub validate_config {
    my ($host, $port) = @_;
    croak "HOST is required" unless $host;
    croak "Invalid port" unless $port > 0 && $port < 65536;
    return 1;
}

File: tests/fixtures/sample.lua
--- @module networking
-- Networking utilities for managing connections.

local json = require("json")
local socket = require("socket")

--- Maximum number of retries.
local MAX_RETRIES = 3

--- Default port for connections.
local DEFAULT_PORT = 8080

--- Logs a message with the given level.
--- @param level string The log level
--- @param message string The message to log
local function log(level, message)
    print(string.format("[%s] %s", level, message))
end

--- Connection class implemented via table.
local Connection = {}
Connection.__index = Connection

--- Creates a new Connection.
--- @param host string The host to connect to
--- @param port number The port number
--- @return Connection
function Connection.new(host, port)
    local self = setmetatable({}, Connection)
    self.host = host
    self.port = port or DEFAULT_PORT
    self.connected = false
    return self
end

--- Connects to the remote host.
function Connection:connect()
    log("INFO", "Connecting to " .. self.host .. ":" .. self.port)
    self.connected = true
    return true
end

--- Disconnects from the remote host.
function Connection:disconnect()
    self.connected = false
end

--- Checks if the connection is active.
function Connection:isConnected()
    return self.connected
end

--- Pool manages multiple connections.
local Pool = {}
Pool.__index = Pool

function Pool.new(host, size)
    local self = setmetatable({}, Pool)
    self.host = host
    self.size = size or 10
    self.connections = {}
    return self
end

function Pool:acquire()
    if #self.connections > 0 then
        return table.remove(self.connections)
    end
    local conn = Connection.new(self.host)
    conn:connect()
    return conn
end

function Pool:release(conn)
    table.insert(self.connections, conn)
end

return {
    Connection = Connection,
    Pool = Pool,
    log = log,
}
